vsp.go

```diff
@@ -57,65 +61,315 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
+	var cancels []func()
+	for {
+		select {
+		case <-ctx.Done():
+			defer outerCancel()
+			fatalMu.Lock()
+			err := fatal
+			fatalMu.Unlock()
+			if err != nil {
+				return err
+			}
+			return ctx.Err()
+		case n := <-c.C:
+			if len(n.AttachedBlocks) == 0 {
+				continue
+			}
+
+			tip := n.AttachedBlocks[len(n.AttachedBlocks)-1]
+			w := vsp.w.Internal()
+
+			// Don't perform any actions while transactions are not synced through
+			// the tip block.
+			rp, err := w.RescanPoint(ctx)
+			if err != nil {
+				return err
+			}
+			if rp != nil {
+				log.Debugf("Skipping autobuyer actions: transactions are not synced on Wallet: %s", vsp.w.Name)
+				continue
+			}
+
+			tipHeader, err := w.BlockHeader(ctx, tip)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			height := int32(tipHeader.Height)
+
+			// Cancel any ongoing ticket purchases which are buying
+			// at an old ticket price or are no longer able to
+			// create mined tickets the window.
+			if height+2 >= nextIntervalStart {
+				for i, cancel := range cancels {
+					cancel()
+					cancels[i] = nil
+				}
+				cancels = cancels[:0]
+
+				intervalSize := int32(w.ChainParams().StakeDiffWindowSize)
+				currentInterval := height / intervalSize
+				nextIntervalStart = (currentInterval + 1) * intervalSize
+
+				// Skip this purchase when no more tickets may be purchased in the interval and
+				// the next sdiff is unknown.  The earliest any ticket may be mined is two
+				// blocks from now, with the next block containing the split transaction
+				// that the ticket purchase spends.
+				if height+2 == nextIntervalStart {
+					log.Debugf("Skipping purchase: next sdiff interval starts soon")
+					continue
+				}
+				// Set expiry to prevent tickets from being mined in the next
+				// sdiff interval.  When the next block begins the new interval,
+				// the ticket is being purchased for the next interval; therefore
+				// increment expiry by a full sdiff window size to prevent it
+				// being mined in the interval after the next.
+				expiry = nextIntervalStart
+				if height+1 == nextIntervalStart {
+					expiry += intervalSize
+				}
+			}
+
+			wal := vsp.w
+			// Get the account balance to determine how many tickets to buy
+			bal, err := wal.GetAccountBalance(int32(vsp.purchaseAccount))
+			if err != nil {
+				return err
+			}
+
+			spendable := bal.Spendable
+			if spendable < balanceToMaintain {
+				log.Debugf("Skipping purchase: low available balance on Wallet: %s", wal.Name)
+				return nil
+			}
+
+			spendable -= balanceToMaintain
+			sdiff, err := wal.Internal().NextStakeDifficultyAfterHeader(ctx, tipHeader)
+			if err != nil {
+				return err
+			}
+
+			buy := int(dcrutil.Amount(spendable) / sdiff)
+			if buy == 0 {
+				log.Debugf("Skipping purchase: low available balance on Wallet: %s", wal.Name)
+				return nil
+			}
+
+			max := int(wal.Internal().ChainParams().MaxFreshStakePerBlock)
+			if buy > max {
+				buy = max
+			}
```

Comment:
Thought the goal was to buy as many as possible, no limits. Instead of reducing the quantity here, a log should do, advising the user that tb is about to purchase more than `max` tickets which means some tickets WILL NOT be included in the next block.

I believe the rationale behind this limiting in dcrwallet is to reduce the possibility of tickets that are purchased now, not making it into the next block. If you look above, you'll notice that ticket purchases are canceled if the next block will cause the ticket price to change and thus any tickets not already mined by then will be rejected (I think).

Probably worth discussing first though, just something to keep in mind for your next commit.